Outgoing reputation and HTLC Accountability

carlaKC opened the pull request:
This PR replaces #1071 with our latest proposal for a reputation scheme and resource management to mitigate slow jamming attacks.

Key differences for the new approach are:
* **Reputation is tracked for outgoing channels**, as only downstream peers are capable of slowing down resolution of htlcs.
* **Resources are limited on incoming channels**, as we're defending against a malicious downstream peer slowing resolution down.
* We now pass an `accountable` signal to our outgoing peer to indicate whether we'll **hold their reputation accountable** because scarce resources have been used (rather than an `endorsed` signal that indicates scarce resources _may_ be used).
* This signal **originates at the receiving node**, as they're ultimately in charge of resolution time, and is **propagated in the onion** to prevent tampering.
* General bucket resources have some DOS protections, and we introduce a congestion bucket to help facilitate honest payments during attack.

carlaKC committed:
11+12: add optional accountable field in invoices

The recipient of a payment is ultimately responsible for the fast
resolution of a payment (though intermediate node can, of course,
slow it down).

We add a signal from the final recipient that the sender can use to
reason about the amount of time it should take to resolve.

carlaKC committed:
11+12: add optional accountable field in invoices

The recipient of a payment is ultimately responsible for the fast
resolution of a payment (though intermediate node can, of course,
slow it down).

We add a signal from the final recipient that the sender can use to
reason about the amount of time it should take to resolve.

Co-authored-by: Thomas HUET <thomas.huet@acinq.fr>

carlaKC committed:
02+04: add accountability signal to update_add_htlc and onion

Once we have a signal from the recipient, we need a way to propagate
this information throughout the route. Including a signal in the onion
provides an uncorruptable way for the sender to propagate this signal.
If the sender is dishonest, this will eventually be detected by the
final recipient.

In the commits that follow we'll add reputation and resource management
that will allow nodes to use this signal to protect against jamming
attacks.

Co-authored-by: Thomas HUET <thomas.huet@acinq.fr>

carlaKC committed:
02+04: add accountability signal to update_add_htlc and onion

Once we have a signal from the recipient, we need a way to propagate
this information throughout the route. Including a signal in the onion
provides an uncorruptable way for the sender to propagate this signal.
If the sender is dishonest, this will eventually be detected by the
final recipient.

In the commits that follow we'll add reputation and resource management
that will allow nodes to use this signal to protect against jamming
attacks.

carlaKC committed:
recommendations: write up full system of reputation

carlaKC commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,516 @@
+
+We define the `opportunity_cost` of the time a HTLC takes to resolve:
+
+`opportunity_cost = ceil ( (resolution_time - resolution_period) / resolution_period) * fees`
```
Updated [here](https://github.com/lightning/bolts/compare/4c35777f44689e825bac0e44cae0150897e9abf4..1bab2ce8ce6403ee580b4b1631898e8a17df06fe).

morehouse reviewed (commented):
Concept ACK.

This is a big improvement over https://github.com/lightning/bolts/pull/1071:

- The inverted reputation scheme protects against [sink attacks](https://delvingbitcoin.org/t/hybrid-jamming-mitigation-results-and-updates/1147#p-3212-manipulation-sink-attack-9).
- The inverted reputation scheme enables new nodes to immediately send payments to reputable destinations when the network is under attack (better UX).
- The hash-based general slot assignment provides some discouragement for general jamming attacks.
- The tit-for-tat congestion bucket provides a mechanism to build reputation even when being general-jammed.

We probably also want to tweak the general slot assignment recommendations a bit to improve the strength of the hash-based defense:

- Scale `general_bucket_slot_count` based on `max_accepted_htlcs` instead of the commitment type.
- Increase the recommended percentage of total slots allocated to the general bucket.

There's also still an open question about how to address the issues with [high payment fanout](https://delvingbitcoin.org/t/hybrid-jamming-mitigation-results-and-updates/1147/4).  For example, LSPs often have high payment fan-out, which would prevent their clients from *ever* getting a good reputation with the current algorithm.

carlaKC commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,529 @@
+- MUST be randomly chosen and unique per channel.
+- SHOULD be persisted across restarts to restore slot allocations.
+
+A HTLC is eligible to use the general bucket if for its
+`(incoming scid, outgoing scid)`'s assigned resources:
+- Currently occupied slots < `general_bucket_slot_count`
+- Currently occupied liquidity + `amt_msat` <= `general_bucket_liquidity_allocation`
```
This is covered by the `general_bucket_slot_count` check above?
It's badly named, I"ll change it to `general_bucket_slot_allocation` so that it's clearer.

carlaKC commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,529 @@
+    average, expressed in seconds.
+- `decaying_average`: stores the value of the decaying average.
+- `decay_rate`: a constant rate of decay based on the rolling window chosen,
+  calculated as: `((1/2)^(2/window_length_seconds))`.
```
`window_length_seconds` is a placeholder for the period that you want a rolling average over - for revenue you'd use `revenue_window` and for reputation you'd use `revenue_window * reputation_multiplier`, for example.

I'll update the wording - let me know if it clarifies it!

morehouse commented on 11-payment-encoding.md:
```diff
@@ -203,6 +204,10 @@ A writer:
   - if a writer offers more than one of any field type, it:
     - MUST specify the most-preferred field first, followed by less-preferred fields, in order.
+  - if the invoice will be resolved within 90 seconds of HTLC(s) arrival:
+    - SHOULD include a `a` field
+  - otherwise:
+    - MUST NOT include a single `a` field
```
Nit
```suggestion
    - SHOULD include a single `a` field
  - otherwise:
    - MUST NOT include an `a` field
```

carlaKC replied:
[fixed!](https://github.com/lightning/bolts/compare/c25f7b11406c6485c9980ec9177ec62c2c35c19c..84ab329edbfbe687f557eb8569f11bad1bf075e7)

morehouse commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,529 @@
+We define the following:
+- `general_bucket_slot_count`:
+  - If the channel type allows a maximum of 483 HTLCs: 20
+  - If the channel type allows a maximum of 120 HTLCs: 5
```
> [LND](https://github.com/lightningnetwork/lnd/blob/f293566849f2fe940aeb03f3b05514590bd54eba/config.go#L184) and [LDK](https://github.com/lightningdevkit/rust-lightning/blob/9514637f8732c66450ec3a5cc2c64c8708d36d67/lightning/src/ln/chan_utils.rs#L69) both have defaults on 483, so most of the network is probably running with this default.

That's true for LND, though it looks like LDK [defaults](https://github.com/lightningdevkit/rust-lightning/blob/81495c6040647573a750eefcafb686edb770ab0d/lightning/src/util/config.rs#L254) to 50.

> Assuming that these values are set to protect nodes from the on-chain cost of resolution, I'm hopeful that other impls will be able to increase the number of slots they allow now that they can't be so trivially filled up.

Good point.  Using the suggested defaults with zero-fee commits would put the general bucket size at `120 * .4 = 48`, which is  essentially the same as current `max_accepted_htlcs` defaults.  Usage above that amount would be possible but would require reputation.

thomash-acinq commented on 02-peer-protocol.md:
```diff
@@ -2233,6 +2233,7 @@ is destined, is described in [BOLT #4](04-onion-routing.md).
     1. type: 0 (`blinded_path`)
     2. data:
         * [`point`:`path_key`]
+    1. type: 1 (`accountable`)
```
Why not allow multiple accountability level instead of it being just a boolean flag? Just as with the endorsement before.

carlaKC committed:
f move accountable into invoice range for bolt-12

carlaKC commented on 12-offer-encoding.md:
```diff
@@ -644,6 +644,7 @@ the `onion_message` `invoice` field.
         * [`name_len*byte`:`name`]
         * [`u8`:`domain_len`]
         * [`domain_len*byte`:`domain`]
+    1. type: 93 (`accountable`)
```
> This must be in the range 160-240

Right - only needs to be in the invoice! Addressed in https://github.com/lightning/bolts/pull/1280/commits/dd99be33aaaeb91c343321e45ddf1d62a1656b76

carlaKC committed:
f: provide precise bucket allocation for zero fee commitment channels

h/t elnosh for pointing this out!

thomash-acinq commented on 04-onion-routing.md:
```diff
@@ -339,6 +347,10 @@ The reader:
       - incoming `amount_msat` < `amt_to_forward`.
```
A few lines above (old line 322: "MUST return an error if the payload contains other tlv fields than"): we need to add `upgrade_accountability` to the list of TLVs allowed.

thomash-acinq committed:
f set `upgrade_accountability` in `encrypted_recipient_data`

Co-authored-by: Thomas HUET <thomas.huet@acinq.fr>

thomash-acinq reviewed (commented):
Passing `upgrade_accountability` in the payload is not compatible with blinded routes, I've fixed it in https://github.com/carlaKC/lightning-rfc/pull/6

carlaKC commented:
> Letting the recipient add upgrade_accountability in the blinded route has another advantage: it allows the HTLC to become accountable inside the blinded route even if the sender does not support accountability.

This seems like the right approach to me, since it's a recipient chosen value. Thanks for the PR, I'll upstream it!

carlaKC committed:
f prefix accoutable field with invoice to align with bolt12 convention

Co-authored-by: Thomas HUET <thomas.huet@acinq.fr>

carlaKC committed:
f typos in reputation algorithm write up

carlaKC committed:
f Add minimum of 5 slots for general bucket allocation

carlaKC committed:
f Use chacha stream instead of sha for indexes

We can just keep pulling from a single stream rather than needing
multiple hashes.

carlaKC committed:
f Update decay rate for decaying average

For a more accurate decaying average, particularly relevant for when
we calculate the average across several windows.

carlaKC committed:
f Add warmup factor to revenue aggregate

Previous approach underestimated in early periods.

carlaKC committed:
f Add note that incredibly small channels can't be supported

carlaKC committed:
f note that we allow protected usage to partially use other buckets

erickcestari commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,551 @@
+This document outlines recommendations for implementing local resource
+conservation to mitigate slow jamming attacks. It is part of a hybrid solution
+to mitigating channel jamming, intended to be followed by the introduction of
+unconditional fees to mitigate fast jamming.
```
nit++: Is quick or fast jamming?

```suggestion
unconditional fees to mitigate quick jamming.
```

erickcestari commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,551 @@
+the resources along the target route in two ways, though the difference
+between the two behaviors is not rigorously defined:
+- Quick Jamming: sending a continuous stream of payments through a route that
+  are quickly failed, sent at a rate that continuously consumes all resources
```
What exactly are resources in this case?

If the HTLC is being held for less than 90 seconds, would this be considered a quick jamming attack?

carlaKC replied:
> What exactly are resources in this case?

The resources are slots + liquidity. Can add clarification to the first sentence of this paragraph if you think it needs it?

> If the HTLC is being held for less than 90 seconds, would this be considered a quick jamming attack?

Yes, but as handwaved in "not rigorously defined" we don't have a strict classification line between the two.

erickcestari replied:
> The resources are slots + liquidity. Can add clarification to the first sentence of this paragraph if you think it needs it?

I think it would be clearer, since the Slow Jamming definition explicitly names the consumed resources.

carlaKC replied:
How's [this](https://github.com/lightning/bolts/compare/2e09bcdfdf0a1670c9df041bd703fd6f9a6e4ecd..b881c0b520ddd89eed08102dd0b0faecb3b52dc0)?

erickcestari replied:
It looks great! Thanks!

erickcestari commented on 11-payment-encoding.md:
```diff
@@ -161,6 +158,7 @@ Currently defined tagged fields are:
 * `9` (5): `data_length` variable. One or more 5-bit values containing features
   supported or required for receiving this payment.
   See [Feature Bits](#feature-bits).
+* `a` (31): `data_length` 0. A marker field that indicates whether the recipient is willing to be held accountable for the timely resolution of the invoice upon HTLC(s) receipt. 
```
According to BIP-0173 (bech32), the number `31` is the letter `l`. To have the letter `a`, we should sign number `29`.

https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki#bech32

```suggestion
* `a` (29): `data_length` 0. A marker field that indicates whether the recipient is willing to be held accountable for the timely resolution of the invoice upon HTLC(s) receipt. 
```

carlaKC replied:
Nice catch!

erickcestari commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,551 @@
+The original sender:
+- SHOULD NOT set `accountable` in their offered `update_add_htlc`.
+- MAY set `accountable` to mimic patterns observed in their forwarded traffic
+  if `upgrade_accountability` is set in the recipient's invoice.
```
nit: The invoice has the `accountable` field.

```suggestion
  if the recipient's invoice set `accountable`.
```

erickcestari commented on 02-peer-protocol.md:
```diff
@@ -2275,6 +2272,18 @@ A sending node:
     - MUST set `path_key` (see [Route Blinding](04-onion-routing.md#route-blinding))
+  - if it is the original source of the HTLC:
+    - SHOULD omit `accountable`.
+    - MAY set `accountable` to mimic patterns of HTLCs it has forwarded.
```
A payer cannot set `accountable` to the HTLC if the invoice does not have the `accountable` field set because the receiver would reject the payment.

>  - If it is the final node:
    - MUST return an error if:
      ...
      - `accountable` is set in the incoming `update_add_htlc` but `accountable` was not set in the invoice.

```suggestion
    - if the recipient's invoice set `accountable`:
      - MAY set `accountable` to mimic patterns of HTLCs it has forwarded.
```

erickcestari commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,551 @@
+## Table of Contents
+
+- [Introduction](#introduction)
+- [Overview](#overview)
```
nit:

```suggestion
- [Overview](#overview)
- [HTLC Accountability](#htlc-accountability)
```

erickcestari commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,551 @@
+- Protected: 40% of `max_accepted_htlcs` and `max_htlc_value_in_flight_msat`.
+
+Note that if a channel has a sufficiently small `max_accepted_htlcs` or
+`max_htlc_value_in_flight_msat`, these buckets cannot provided meaningful
```
nit:
```suggestion
`max_htlc_value_in_flight_msat`, these buckets cannot provide meaningful
```

erickcestari commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,551 @@
+#### Rationale
+
+The protected bucket aims to project unrestricted resource access to peers that
+have build sufficient reputation. In the unlikely event that the protected
```
nit:
```suggestion
have built sufficient reputation. In the unlikely event that the protected
```

erickcestari commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,551 @@
+
+In times of peace when general resources are not saturated, there is no need
+for nodes to hold any HTLCs `accountable`. When resources become scarce, it is
+only safe to allow nodes that have build reputation to access protected
```
nit:
```suggestion
only safe to allow nodes that have built reputation to access protected
```

erickcestari commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,551 @@
+  Behavior of downstream peers is observed because a malicious downstream peer
+  is necessary to perform a slow-jamming attack (the upstream may or may not be
+  malicious).
+- [HTLC accountability](#../02-peer-protocol.md#adding-an-htlc-update_add_htlc):
```
nit:
```suggestion
- [HTLC accountability](../02-peer-protocol.md#adding-an-htlc-update_add_htlc):
```

erickcestari commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,551 @@
+
+The forwarding node:
+- If `accountable` is set in the received `update_add_htlc`:
+  - MUST set `accountable` in the offered `update_add_htlc` (see [BOLT-02](#../02-peer-protocol.md#adding-an-htlc-update_add_htlc)).
```
nit:
```suggestion
  - MUST set `accountable` in the offered `update_add_htlc` (see [BOLT-02](../02-peer-protocol.md#adding-an-htlc-update_add_htlc)).
```

erickcestari reviewed (commented):
Did a pass over the recommendation doc and the BOLT 02/04/11/12 changes, design reads well and the permission (BOLT-02) vs. policy (recommendation) split is coherent.

One real issue: the BOLT-11 `a` field is numbered `31`, but `31` is `l` in `bech32`; `a` is `29`. Worth fixing before merge.

Everything else is nits: typos, broken cross-file links (leading #), missing TOC entry, and tightening the sender-mimicry wording.

carlaKC committed:
recommendations: write up full system of reputation

erickcestari commented on recommendations/local-resource-conservation.md:
```diff
@@ -0,0 +1,552 @@
+- `last_update`: stores the timestamp of the last update to the decaying
+    average, expressed in seconds.
+- `decaying_average`: stores the value of the decaying average.
+- `decay_rate`: a constant rate of decay based on the rolling window chosen,
+  calculated as: `(1/2)^(1/ln(2)*window_length_seconds)`
```
I think the `decay_rate` formula in the spec is missing a parenthesis: `(1/2)^(1/ln(2)*window_length_seconds)` as written parses as `0.5^(W/ln 2) = e^(-W)`, which is essentially zero for any realistic window. I guess the intended formula was `(1/2)^(1/(ln(2)*window_length_seconds))`, which simplifies to `e^(-1/W)`.

One thing I noticed is that LND and LDK are currently using `0.5^(2/W)` (half-life = W/2), which decays a bit faster than `e^(-1/W)`. With the 24-week window, `0.5^(2/W)` leaves 25% weight at the end vs 37% for `e^(-1/W)`.

So which one should be the recommended `decay_rate` here? Should the spec just get the parenthesis fix and keep `e^(-1/W)`, or would it make sense to align it with what the implementations are already doing, i.e. `(1/2)^(2/window_length_seconds)`?

```suggestion
- `decay_rate`: a constant rate of decay based on the rolling window chosen,
  calculated as: `(1/2)^(1/(ln(2)*window_length_seconds))`
```
